{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Use this as a reference to pull data from LIMS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pg8000          #pg8000 access SQL databases\n",
    "import pandas as pd    #pandas will be needed to work in a dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#code from Agata\n",
    "#these are nice functions to open LIMS, make a query and then close LIMS after\n",
    "\n",
    "def _connect(user=\"limsreader\", host=\"limsdb2\", database=\"lims2\", password=\"limsro\", port=5432):\n",
    "    conn = pg8000.connect(user=user, host=host, database=database, password=password, port=port)\n",
    "    return conn, conn.cursor()\n",
    "\n",
    "def _select(cursor, query):\n",
    "    cursor.execute(query)\n",
    "    columns = [ d[0] for d in cursor.description ]\n",
    "    return [ dict(zip(columns, c)) for c in cursor.fetchall() ]\n",
    "\n",
    "def limsquery(query, user=\"limsreader\", host=\"limsdb2\", database=\"lims2\", password=\"limsro\", port=5432):\n",
    "    \"\"\"A function that takes a string containing a SQL query, connects to the LIMS database and outputs the result.\"\"\"\n",
    "    conn, cursor = _connect(user, host, database, password, port)\n",
    "    try:\n",
    "        results = _select(cursor, query)\n",
    "    finally:\n",
    "        \n",
    "        #THESE ARE IMPORTANT!!!!!!\n",
    "        #Every query needs to be closed when done\n",
    "        cursor.close()             \n",
    "        conn.close()\n",
    "    return results\n",
    "\n",
    "\n",
    "#this last function will take our query results and put them in a dataframe so that they are easy to work with\n",
    "def get_lims_dataframe(query):\n",
    "    '''Return a dataframe with lims query'''\n",
    "    result = limsquery(query)\n",
    "    try:\n",
    "        data_df = pd.DataFrame(data=result, columns=result[0].keys())\n",
    "    except IndexError:\n",
    "        print \"Could not find results for your query.\"\n",
    "        data_df = pd.DataFrame()\n",
    "    return data_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'cell_depth': None, 'ephys_roi_result_id': None, 'parent_y_coord': 0, 'reference_space_id': None, 'updated_at': datetime.datetime(2016, 12, 16, 4, 54, 44, 477335), 'cell_label': None, 'preparation_method_id': None, 'parent_x_coord': 2, 'location_id': None, 'id': 556516441, 'cortex_layer_id': None, 'plane_of_section_id': 11, 'frozen_at': None, 'flipped_specimen_id': 561557765, 'data': None, 'pinned_radius': None, 'rna_integrity_number': None, 'histology_well_name': None, 'created_by': None, 'priority': None, 'parent_id': 556516212, 'ephys_start_time_sec': None, 'project_id': 305094322, 'alignment3d_id': None, 'carousel_well_name': u'T301_122_161107_01_12', 'patched_cell_container': None, 'updated_by': None, 'cell_prep_id': None, 'biophysical_model_state': u'review_required', 'barcode': u'0556516441', 'storage_directory': None, 'x_coord': None, 'tissue_ph': None, 'specimen_preparation_method_id': None, 'donor_id': 555257198, 'operation_id': None, 'ephys_neural_tissue_plan_id': 555257244, 'structure_id': None, 'parent_z_coord': 0, 'facs_well_id': None, 'name': u'Ndnf-IRES2-dgCre;Ai14-280612.06.02', 'y_coord': None, 'normalization_group_id': None, 'postmortem_interval_id': None, 'specimen_set_id': None, 'created_at': datetime.datetime(2016, 11, 7, 16, 3, 11, 459275), 'tissue_processing_id': 555257241, 'ephys_cell_plan_id': None, 'hemisphere_id': None, 'ephys_qc_result': None, 'cell_reporter_id': None, 'task_flow_id': None, 'external_specimen_name': None}\n"
     ]
    }
   ],
   "source": [
    "my_query = \"SELECT * FROM specimens LIMIT 10\"\n",
    "#we are going to select all columns in the specimens table and limit our search to the first 10\n",
    "\n",
    "my_result = limsquery(my_query)\n",
    "\n",
    "first_element = my_result[0]\n",
    "#now we only want to look at the first element of our result\n",
    "\n",
    "print first_element"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['weight', 'education_level_id', 'updated_at', 'height', 'transgenic_induction_method_id', 'id', 'death_on', 'primary_tissue_source_id', 'full_genotype', 'induction_method', 'age_id', 'created_by', 'occupation_id', 'date_of_birth', 'external_donor_name', 'race_id', 'death_manner_id', 'updated_by', 'organism_id', 'data', 'name', 'created_at', 'gender_id', 'baseline_weight_g', 'death_cause_id', 'handedness_id']\n"
     ]
    }
   ],
   "source": [
    "my_query = \"SELECT * FROM donors LIMIT 10\"\n",
    "#we are going to select all columns in the specimens table and limit our search to the first 10\n",
    "\n",
    "my_result = limsquery(my_query)\n",
    "\n",
    "first_element = my_result[0]\n",
    "#now we only want to look at the first element of our result\n",
    "\n",
    "print first_element.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['published_at', 'workflow_state', 'created_at', 'updated_at', 'filename', 'storage_directory', 'file_source_id', 'attachable_type', 'content_type', 'well_known_file_type_id', 'attachable_id', 'id', 'size']\n"
     ]
    }
   ],
   "source": [
    "my_query = \"SELECT * FROM well_known_files LIMIT 10\"\n",
    "my_result = limsquery(my_query)\n",
    "first_element = my_result[0]\n",
    "print first_element.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'table_name': u'pg_type'},\n",
       " {'table_name': u'pg_roles'},\n",
       " {'table_name': u'pg_group'},\n",
       " {'table_name': u'pg_user'},\n",
       " {'table_name': u'pg_rules'},\n",
       " {'table_name': u'pg_views'},\n",
       " {'table_name': u'pg_tables'},\n",
       " {'table_name': u'pg_matviews'},\n",
       " {'table_name': u'pg_indexes'},\n",
       " {'table_name': u'pg_stats'},\n",
       " {'table_name': u'pg_settings'},\n",
       " {'table_name': u'pg_locks'},\n",
       " {'table_name': u'pg_cursors'},\n",
       " {'table_name': u'pg_available_extensions'},\n",
       " {'table_name': u'pg_available_extension_versions'},\n",
       " {'table_name': u'pg_prepared_xacts'},\n",
       " {'table_name': u'pg_prepared_statements'},\n",
       " {'table_name': u'pg_seclabels'},\n",
       " {'table_name': u'pg_timezone_abbrevs'},\n",
       " {'table_name': u'pg_timezone_names'},\n",
       " {'table_name': u'pg_stat_all_tables'},\n",
       " {'table_name': u'pg_stat_xact_all_tables'},\n",
       " {'table_name': u'pg_stat_sys_tables'},\n",
       " {'table_name': u'pg_stat_xact_sys_tables'},\n",
       " {'table_name': u'pg_stat_user_tables'},\n",
       " {'table_name': u'pg_stat_xact_user_tables'},\n",
       " {'table_name': u'pg_statio_all_tables'},\n",
       " {'table_name': u'pg_statio_sys_tables'},\n",
       " {'table_name': u'pg_statio_user_tables'},\n",
       " {'table_name': u'pg_statio_all_indexes'},\n",
       " {'table_name': u'pg_statio_sys_indexes'},\n",
       " {'table_name': u'pg_statio_user_indexes'},\n",
       " {'table_name': u'pg_statio_all_sequences'},\n",
       " {'table_name': u'pg_statio_sys_sequences'},\n",
       " {'table_name': u'pg_statio_user_sequences'},\n",
       " {'table_name': u'pg_stat_activity'},\n",
       " {'table_name': u'pg_stat_replication'},\n",
       " {'table_name': u'pg_replication_slots'},\n",
       " {'table_name': u'pg_stat_database'},\n",
       " {'table_name': u'pg_stat_database_conflicts'},\n",
       " {'table_name': u'pg_stat_user_functions'},\n",
       " {'table_name': u'pg_stat_xact_user_functions'},\n",
       " {'table_name': u'pg_stat_archiver'},\n",
       " {'table_name': u'pg_stat_bgwriter'},\n",
       " {'table_name': u'pg_user_mappings'},\n",
       " {'table_name': u'pg_stat_all_indexes'},\n",
       " {'table_name': u'pg_stat_sys_indexes'},\n",
       " {'table_name': u'pg_stat_user_indexes'},\n",
       " {'table_name': u'information_schema_catalog_name'},\n",
       " {'table_name': u'applicable_roles'},\n",
       " {'table_name': u'administrable_role_authorizations'},\n",
       " {'table_name': u'attributes'},\n",
       " {'table_name': u'character_sets'},\n",
       " {'table_name': u'check_constraint_routine_usage'},\n",
       " {'table_name': u'check_constraints'},\n",
       " {'table_name': u'collations'},\n",
       " {'table_name': u'collation_character_set_applicability'},\n",
       " {'table_name': u'column_domain_usage'},\n",
       " {'table_name': u'column_privileges'},\n",
       " {'table_name': u'column_udt_usage'},\n",
       " {'table_name': u'columns'},\n",
       " {'table_name': u'constraint_column_usage'},\n",
       " {'table_name': u'constraint_table_usage'},\n",
       " {'table_name': u'domain_constraints'},\n",
       " {'table_name': u'domain_udt_usage'},\n",
       " {'table_name': u'domains'},\n",
       " {'table_name': u'enabled_roles'},\n",
       " {'table_name': u'key_column_usage'},\n",
       " {'table_name': u'parameters'},\n",
       " {'table_name': u'referential_constraints'},\n",
       " {'table_name': u'role_column_grants'},\n",
       " {'table_name': u'routine_privileges'},\n",
       " {'table_name': u'role_routine_grants'},\n",
       " {'table_name': u'routines'},\n",
       " {'table_name': u'schemata'},\n",
       " {'table_name': u'sequences'},\n",
       " {'table_name': u'pg_attribute'},\n",
       " {'table_name': u'pg_proc'},\n",
       " {'table_name': u'table_constraints'},\n",
       " {'table_name': u'table_privileges'},\n",
       " {'table_name': u'role_table_grants'},\n",
       " {'table_name': u'tables'},\n",
       " {'table_name': u'triggered_update_columns'},\n",
       " {'table_name': u'triggers'},\n",
       " {'table_name': u'udt_privileges'},\n",
       " {'table_name': u'role_udt_grants'},\n",
       " {'table_name': u'usage_privileges'},\n",
       " {'table_name': u'role_usage_grants'},\n",
       " {'table_name': u'user_defined_types'},\n",
       " {'table_name': u'view_column_usage'},\n",
       " {'table_name': u'view_routine_usage'},\n",
       " {'table_name': u'view_table_usage'},\n",
       " {'table_name': u'views'},\n",
       " {'table_name': u'data_type_privileges'},\n",
       " {'table_name': u'element_types'},\n",
       " {'table_name': u'column_options'},\n",
       " {'table_name': u'foreign_data_wrapper_options'},\n",
       " {'table_name': u'foreign_data_wrappers'},\n",
       " {'table_name': u'foreign_server_options'},\n",
       " {'table_name': u'foreign_servers'},\n",
       " {'table_name': u'foreign_table_options'},\n",
       " {'table_name': u'foreign_tables'},\n",
       " {'table_name': u'user_mapping_options'},\n",
       " {'table_name': u'user_mappings'},\n",
       " {'table_name': u'pg_constraint'},\n",
       " {'table_name': u'pg_index'},\n",
       " {'table_name': u'pg_operator'},\n",
       " {'table_name': u'pg_opfamily'},\n",
       " {'table_name': u'pg_opclass'},\n",
       " {'table_name': u'pg_am'},\n",
       " {'table_name': u'ephys_qc_criterias'},\n",
       " {'table_name': u'rgb_values'},\n",
       " {'table_name': u'refseq_human_gff_37'},\n",
       " {'table_name': u'rseq_experiments'},\n",
       " {'table_name': u'cell_prep_samples_specimens'},\n",
       " {'table_name': u'rseq_tube_inputs'},\n",
       " {'table_name': u'rseq_tubes'},\n",
       " {'table_name': u'analysis_run_containers'},\n",
       " {'table_name': u'avg_transformations'},\n",
       " {'table_name': u'executables'},\n",
       " {'table_name': u'ophys_experiment_analysis_runs'},\n",
       " {'table_name': u'rseq_experiment_components_star_analysis_runs'},\n",
       " {'table_name': u'methods_versions_task_templates'},\n",
       " {'table_name': u'transcripts'},\n",
       " {'table_name': u'pg_depend'},\n",
       " {'table_name': u'flipped_specimens'},\n",
       " {'table_name': u'wij_sets'},\n",
       " {'table_name': u'avg_graphic_objects'},\n",
       " {'table_name': u'task_templates'},\n",
       " {'table_name': u'avg_microarray_templates'},\n",
       " {'table_name': u'job_queue_notification_field_values'},\n",
       " {'table_name': u'loci'},\n",
       " {'table_name': u'sessions'},\n",
       " {'table_name': u'image_structure_data'},\n",
       " {'table_name': u'email_aliases_projects'},\n",
       " {'table_name': u'sub_images'},\n",
       " {'table_name': u'genes'},\n",
       " {'table_name': u'job_queue_notifications_users'},\n",
       " {'table_name': u'rseq_library_preps'},\n",
       " {'table_name': u'sql_languages'},\n",
       " {'table_name': u'sql_packages'},\n",
       " {'table_name': u'sql_sizing'},\n",
       " {'table_name': u'sql_sizing_profiles'},\n",
       " {'table_name': u'ephys_roi_results'},\n",
       " {'table_name': u'genomic_reference_updates'},\n",
       " {'table_name': u'differentiation_designs'},\n",
       " {'table_name': u'wij_chunks'},\n",
       " {'table_name': u'graph_views'},\n",
       " {'table_name': u'differentiation_designs_differentiation_stages'},\n",
       " {'table_name': u'handednesses'},\n",
       " {'table_name': u'cell_prep_sample_types'},\n",
       " {'table_name': u'annovaluemaps'},\n",
       " {'table_name': u'primers'},\n",
       " {'table_name': u'refseq_macaque_gff'},\n",
       " {'table_name': u'well_known_file_types'},\n",
       " {'table_name': u'annotationdetail'},\n",
       " {'table_name': u'differentiation_sets'},\n",
       " {'table_name': u'orphanedannotationsets'},\n",
       " {'table_name': u'probe_alignments'},\n",
       " {'table_name': u'postupdate'},\n",
       " {'table_name': u'orphanedentities'},\n",
       " {'table_name': u'email_aliases'},\n",
       " {'table_name': u'differentiation_stage_instances'},\n",
       " {'table_name': u'donors_genotypes'},\n",
       " {'table_name': u'equipment_runplans'},\n",
       " {'table_name': u'probe_expressed_states'},\n",
       " {'table_name': u'genotype_types'},\n",
       " {'table_name': u'planregionmaps'},\n",
       " {'table_name': u'planregionmapcells'},\n",
       " {'table_name': u'linkannotation'},\n",
       " {'table_name': u'cell_roi_analysis_runs'},\n",
       " {'table_name': u'roi_exclusion_training_labels'},\n",
       " {'table_name': u'analysis_runs_types'},\n",
       " {'table_name': u'annotated_regions'},\n",
       " {'table_name': u'api_column_definitions'},\n",
       " {'table_name': u'cell_rois_roi_exclusion_labels'},\n",
       " {'table_name': u'cell_rois_roi_exclusion_training_labels'},\n",
       " {'table_name': u'ercc_sets'},\n",
       " {'table_name': u'flat_field_correction_runs'},\n",
       " {'table_name': u'organisms'},\n",
       " {'table_name': u'project_sets'},\n",
       " {'table_name': u'reference_to_reference_transforms'},\n",
       " {'table_name': u'refseq_mouse_gff'},\n",
       " {'table_name': u'project_sets_projects'},\n",
       " {'table_name': u'reference3d_spaces'},\n",
       " {'table_name': u'roi_exclusion_labels'},\n",
       " {'table_name': u'ext_genes_image_series_v'},\n",
       " {'table_name': u'tissue_processings'},\n",
       " {'table_name': u'behavior_time_blocks'},\n",
       " {'table_name': u'blast_versions'},\n",
       " {'table_name': u'specimen_hierarchies'},\n",
       " {'table_name': u'experiment_sessions'},\n",
       " {'table_name': u'cell_prep_samples'},\n",
       " {'table_name': u'ext_genes_v'},\n",
       " {'table_name': u'refseq_human_gff'},\n",
       " {'table_name': u'ext_image_series_treatments_v'},\n",
       " {'table_name': u'rseq_experiment_components'},\n",
       " {'table_name': u'specimen_blocks'},\n",
       " {'table_name': u'scheduled_procedures_subjects'},\n",
       " {'table_name': u'southern_blots'},\n",
       " {'table_name': u'spatial_mappings'},\n",
       " {'table_name': u'affymetrix_probesets'},\n",
       " {'table_name': u'cell_additives'},\n",
       " {'table_name': u'cell_prep_roi_plans_cell_prep_templates'},\n",
       " {'table_name': u'calculatedilutions_v'},\n",
       " {'table_name': u'hemispheres'},\n",
       " {'table_name': u'graph_views_structures'},\n",
       " {'table_name': u'cdna_primer_report_views'},\n",
       " {'table_name': u'methods_versions_watchers'},\n",
       " {'table_name': u'job_queue_notification_reports_job_queue_notifications'},\n",
       " {'table_name': u'polygon_spatial_mappings'},\n",
       " {'table_name': u'injection_materials_surgical_procedures'},\n",
       " {'table_name': u'cel1_nucleases'},\n",
       " {'table_name': u'job_queue_notifications_job_queues'},\n",
       " {'table_name': u'surgical_procedures'},\n",
       " {'table_name': u'specimen_metadata'},\n",
       " {'table_name': u'tasks_task_templates'},\n",
       " {'table_name': u'analysis_runs'},\n",
       " {'table_name': u'job_queue_notifications'},\n",
       " {'table_name': u'job_queue_notification_field_values_job_queue_notifications'},\n",
       " {'table_name': u'job_queue_notification_criterias_job_queue_notifications'},\n",
       " {'table_name': u'culture_conditions'},\n",
       " {'table_name': u'mouse_chrs_map'},\n",
       " {'table_name': u'macaque_chrs_map'},\n",
       " {'table_name': u'neuron_reconstruction_tags'},\n",
       " {'table_name': u'color_channel_descriptions'},\n",
       " {'table_name': u'cultured_cells'},\n",
       " {'table_name': u'methods_versions'},\n",
       " {'table_name': u'colors'},\n",
       " {'table_name': u'log_messages'},\n",
       " {'table_name': u'ncbi_sequence_statuses'},\n",
       " {'table_name': u'operations'},\n",
       " {'table_name': u'probe_pcr_products'},\n",
       " {'table_name': u'reference_spaces'},\n",
       " {'table_name': u'clone_primer_report_views'},\n",
       " {'table_name': u'cell_preps'},\n",
       " {'table_name': u'methods_versions_ncrs'},\n",
       " {'table_name': u'operation_locations_operations'},\n",
       " {'table_name': u'probe_refseq_blast_results'},\n",
       " {'table_name': u'cell_preps_cell_prep_roi_plans'},\n",
       " {'table_name': u'visual_behavior_details'},\n",
       " {'table_name': u'injection_material_propagation_directions'},\n",
       " {'table_name': u'donor_dna_constructs_engineering_steps'},\n",
       " {'table_name': u'donor_dna_construct_designs_donor_dna_promoters'},\n",
       " {'table_name': u'ephys_neural_tissue_plans'},\n",
       " {'table_name': u'differentiation_stages_task_flows'},\n",
       " {'table_name': u'facs'},\n",
       " {'table_name': u'cell_rois'},\n",
       " {'table_name': u'ophys_cell_segmentation_runs'},\n",
       " {'table_name': u'engineering_drug_instances'},\n",
       " {'table_name': u'differentiation_stages'},\n",
       " {'table_name': u'facs_population_plans_facs'},\n",
       " {'table_name': u'donor_dna_construct_types'},\n",
       " {'table_name': u'ephys_roi_plans'},\n",
       " {'table_name': u'donor_dna_connections'},\n",
       " {'table_name': u'specimen_sets_treatments'},\n",
       " {'table_name': u'animal_monitoring_details'},\n",
       " {'table_name': u'ncrs_specimens'},\n",
       " {'table_name': u'email_aliases_methods_version_types'},\n",
       " {'table_name': u'stem_cell_line_states'},\n",
       " {'table_name': u'specimen_types_specimens'},\n",
       " {'table_name': u'ephys_stimuli'},\n",
       " {'table_name': u'annotated_region_histories'},\n",
       " {'table_name': u'cross_ontology_mappings'},\n",
       " {'table_name': u'bam_anonymization_runs'},\n",
       " {'table_name': u'specimen_tags'},\n",
       " {'table_name': u'runplan_states'},\n",
       " {'table_name': u'specimens_workflows'},\n",
       " {'table_name': u'ephys_raw_stimulus_names'},\n",
       " {'table_name': u'ephys_sweep_tags_ephys_sweeps'},\n",
       " {'table_name': u'ephys_neural_tissue_plans_ephys_roi_plans'},\n",
       " {'table_name': u'executables_run_types'},\n",
       " {'table_name': u'blocks'},\n",
       " {'table_name': u'scheduled_procedure_notifications_users'},\n",
       " {'table_name': u'blast_db_types'},\n",
       " {'table_name': u'ext_genelookups_v'},\n",
       " {'table_name': u'ephys_sweeps'},\n",
       " {'table_name': u'primary_tissue_sources'},\n",
       " {'table_name': u'postmortem_intervals'},\n",
       " {'table_name': u'executables_run_group_types'},\n",
       " {'table_name': u'executables_workflows'},\n",
       " {'table_name': u'avg_group_labels'},\n",
       " {'table_name': u'cell_prep_sample_collection_types'},\n",
       " {'table_name': u'ercc_stock_pool_components'},\n",
       " {'table_name': u'notification_types'},\n",
       " {'table_name': u'ext_structures_v'},\n",
       " {'table_name': u'specimens'},\n",
       " {'table_name': u'facs_sops'},\n",
       " {'table_name': u'flat_structures_v'},\n",
       " {'table_name': u'notification_criteria_types'},\n",
       " {'table_name': u'operation_types'},\n",
       " {'table_name': u'scheduled_procedure_notification_date_ranges_scheduled_procedur'},\n",
       " {'table_name': u'notification_run_frequencies'},\n",
       " {'table_name': u'fine_structure_searches'},\n",
       " {'table_name': u'ncrs'},\n",
       " {'table_name': u'facs_population_plans'},\n",
       " {'table_name': u'gene_tags_genes'},\n",
       " {'table_name': u'isi_experiments'},\n",
       " {'table_name': u'operation_locations'},\n",
       " {'table_name': u'go_exons'},\n",
       " {'table_name': u'gene_classifications_genes'},\n",
       " {'table_name': u'occupations'},\n",
       " {'table_name': u'notification_reports'},\n",
       " {'table_name': u'genomic_objects'},\n",
       " {'table_name': u'scheduled_procedure_notification_criterias_scheduled_procedure_'},\n",
       " {'table_name': u'assigned_users_scheduled_procedure_notifications'},\n",
       " {'table_name': u'scheduled_procedure_notification_reports_scheduled_procedure_no'},\n",
       " {'table_name': u'gene_termini'},\n",
       " {'table_name': u'notification_criterias'},\n",
       " {'table_name': u'projects_scheduled_procedure_notifications'},\n",
       " {'table_name': u'unit_of_times'},\n",
       " {'table_name': u'injection_coordinates'},\n",
       " {'table_name': u'informatics_ranking_queries'},\n",
       " {'table_name': u'scheduled_procedure_notifications'},\n",
       " {'table_name': u'scheduled_procedure_notification_date_ranges'},\n",
       " {'table_name': u'injection_materials'},\n",
       " {'table_name': u'unit_of_measurements'},\n",
       " {'table_name': u'informatics_ranking_query_results'},\n",
       " {'table_name': u'scheduled_procedure_notification_requested_date_ranges_schedule'},\n",
       " {'table_name': u'facs_well_templates'},\n",
       " {'table_name': u'job_transitions_scanners'},\n",
       " {'table_name': u'job_transitions_workflows'},\n",
       " {'table_name': u'donor_substances'},\n",
       " {'table_name': u'homology_arms'},\n",
       " {'table_name': u'substances'},\n",
       " {'table_name': u'equipment'},\n",
       " {'table_name': u'job_transitions_run_group_types'},\n",
       " {'table_name': u'pcr_polymerases'},\n",
       " {'table_name': u'ophys_experiment_tags'},\n",
       " {'table_name': u'eye_trackings'},\n",
       " {'table_name': u'ophys_sessions'},\n",
       " {'table_name': u'imaging_depths'},\n",
       " {'table_name': u'originating_labs'},\n",
       " {'table_name': u'nuclease_activities_pcrs'},\n",
       " {'table_name': u'probe_summary_views'},\n",
       " {'table_name': u'projects_trigger_types'},\n",
       " {'table_name': u'project_well_types'},\n",
       " {'table_name': u'experiment_containers'},\n",
       " {'table_name': u'reference_spaces_workflows'},\n",
       " {'table_name': u'projection_structure_unionizes'},\n",
       " {'table_name': u'probe_refseq_blast_results_v'},\n",
       " {'table_name': u'related_jobs_v'},\n",
       " {'table_name': u'rseq_library_quantification_methods'},\n",
       " {'table_name': u'images'},\n",
       " {'table_name': u'ophys_experiments'},\n",
       " {'table_name': u'rseq_library_prep_sets'},\n",
       " {'table_name': u'sections_slides_mappings'},\n",
       " {'table_name': u'methods_versions_reviewers'},\n",
       " {'table_name': u'medical_condition_types'},\n",
       " {'table_name': u'sequencing_instruments'},\n",
       " {'table_name': u'scheduled_procedures'},\n",
       " {'table_name': u'methods_version_types'},\n",
       " {'table_name': u'slide_registration_states'},\n",
       " {'table_name': u'specimen_tags_specimens'},\n",
       " {'table_name': u'stem_cell_line_sources'},\n",
       " {'table_name': u'stem_cell_line_sources_stem_cell_lines'},\n",
       " {'table_name': u'structure_sets'},\n",
       " {'table_name': u'task_templates_surgical_procedures'},\n",
       " {'table_name': u'tcp_sops'},\n",
       " {'table_name': u'talen_pairs'},\n",
       " {'table_name': u'task_flows'},\n",
       " {'table_name': u'transfection_condition_colony_colors'},\n",
       " {'table_name': u'trigger_types'},\n",
       " {'table_name': u'user_header_elements'},\n",
       " {'table_name': u'view_slide_parent_specimens'},\n",
       " {'table_name': u'heatmap_runs'},\n",
       " {'table_name': u'ephys_cell_plans'},\n",
       " {'table_name': u'rseq_experiment_components_single_cell_analysis_runs'},\n",
       " {'table_name': u'facs_facs_plate_templates'},\n",
       " {'table_name': u'gene_associations'},\n",
       " {'table_name': u'annotationset'},\n",
       " {'table_name': u'classificationsynonym'},\n",
       " {'table_name': u'allseries'},\n",
       " {'table_name': u'structures'},\n",
       " {'table_name': u'dataset'},\n",
       " {'table_name': u'lims2imageseries'},\n",
       " {'table_name': u'measuregroupvaluedefinition'},\n",
       " {'table_name': u'point'},\n",
       " {'table_name': u'class_mappings'},\n",
       " {'table_name': u'clone_sources'},\n",
       " {'table_name': u'clones'},\n",
       " {'table_name': u'probe_fluor_combinations'},\n",
       " {'table_name': u'probe_sets_run_groups'},\n",
       " {'table_name': u'donors'},\n",
       " {'table_name': u'ensembl_genes_v'},\n",
       " {'table_name': u'ext_image_series_v'},\n",
       " {'table_name': u'fluor_colors'},\n",
       " {'table_name': u'freezers'},\n",
       " {'table_name': u'ncbi_genes_v'},\n",
       " {'table_name': u'refseq_human_gff_36'},\n",
       " {'table_name': u'run_group_states'},\n",
       " {'table_name': u'scanners'},\n",
       " {'table_name': u'pg_event_trigger'},\n",
       " {'table_name': u'users_work_groups'},\n",
       " {'table_name': u'widgets'},\n",
       " {'table_name': u'units'},\n",
       " {'table_name': u'thicknesses'},\n",
       " {'table_name': u'view_calculate_dilutions'},\n",
       " {'table_name': u'work_groups'},\n",
       " {'table_name': u'transfection_conditions'},\n",
       " {'table_name': u'users'},\n",
       " {'table_name': u'well_known_files'},\n",
       " {'table_name': u'pg_attrdef'},\n",
       " {'table_name': u'pg_amop'},\n",
       " {'table_name': u'pg_amproc'},\n",
       " {'table_name': u'pg_language'},\n",
       " {'table_name': u'pg_aggregate'},\n",
       " {'table_name': u'pg_rewrite'},\n",
       " {'table_name': u'pg_trigger'},\n",
       " {'table_name': u'treatment_tasks'},\n",
       " {'table_name': u'pg_description'},\n",
       " {'table_name': u'pg_cast'},\n",
       " {'table_name': u'pg_enum'},\n",
       " {'table_name': u'pg_namespace'},\n",
       " {'table_name': u'pg_conversion'},\n",
       " {'table_name': u'pg_database'},\n",
       " {'table_name': u'pg_db_role_setting'},\n",
       " {'table_name': u'pg_tablespace'},\n",
       " {'table_name': u'pg_pltemplate'},\n",
       " {'table_name': u'pg_auth_members'},\n",
       " {'table_name': u'pg_shdepend'},\n",
       " {'table_name': u'pg_shdescription'},\n",
       " {'table_name': u'pg_ts_config'},\n",
       " {'table_name': u'pg_ts_config_map'},\n",
       " {'table_name': u'pg_ts_dict'},\n",
       " {'table_name': u'pg_ts_parser'},\n",
       " {'table_name': u'pg_ts_template'},\n",
       " {'table_name': u'pg_extension'},\n",
       " {'table_name': u'pg_foreign_data_wrapper'},\n",
       " {'table_name': u'pg_foreign_server'},\n",
       " {'table_name': u'pg_foreign_table'},\n",
       " {'table_name': u'pg_default_acl'},\n",
       " {'table_name': u'pg_seclabel'},\n",
       " {'table_name': u'pg_shseclabel'},\n",
       " {'table_name': u'pg_collation'},\n",
       " {'table_name': u'pg_range'},\n",
       " {'table_name': u'sql_implementation_info'},\n",
       " {'table_name': u'pg_class'},\n",
       " {'table_name': u'pg_inherits'},\n",
       " {'table_name': u'sql_features'},\n",
       " {'table_name': u'pg_largeobject_metadata'},\n",
       " {'table_name': u'annotationplan'},\n",
       " {'table_name': u'alignment2ds'},\n",
       " {'table_name': u'annotationplanreport'},\n",
       " {'table_name': u'stringannotation'},\n",
       " {'table_name': u'preupdate'},\n",
       " {'table_name': u'postupdateall'},\n",
       " {'table_name': u'preupdateall'},\n",
       " {'table_name': u'affymetrix_transcript_clusters'},\n",
       " {'table_name': u'alignment3ds'},\n",
       " {'table_name': u'targetentity'},\n",
       " {'table_name': u'annotationplanuser'},\n",
       " {'table_name': u'planregionmap'},\n",
       " {'table_name': u'measuregroup'},\n",
       " {'table_name': u'groupuser'},\n",
       " {'table_name': u'reflabelupdates'},\n",
       " {'table_name': u'atlases_avg_group_labels'},\n",
       " {'table_name': u'behavior_protocols'},\n",
       " {'table_name': u'activity_cells'},\n",
       " {'table_name': u'atlases'},\n",
       " {'table_name': u'buffers'},\n",
       " {'table_name': u'array_slides'},\n",
       " {'table_name': u'affymetrix_macaque_probesets'},\n",
       " {'table_name': u'cell_descriptions'},\n",
       " {'table_name': u'avg_replicate_types'},\n",
       " {'table_name': u'behavior_trainings'},\n",
       " {'table_name': u'cell_prep_roi_plans_chemicals'},\n",
       " {'table_name': u'crisprs'},\n",
       " {'table_name': u'blast_db_files'},\n",
       " {'table_name': u'crisprs_nuclease_activities'},\n",
       " {'table_name': u'probe_templates'},\n",
       " {'table_name': u'sequences'},\n",
       " {'table_name': u'custom_reports'},\n",
       " {'table_name': u'chemicals'},\n",
       " {'table_name': u'cell_bank_types'},\n",
       " {'table_name': u'crisprs_engineering_steps'},\n",
       " {'table_name': u'biocytin_wells'},\n",
       " {'table_name': u'cell_prep_roi_plans'},\n",
       " {'table_name': u'behavior_sessions'},\n",
       " {'table_name': u'engineering_drugs'},\n",
       " {'table_name': u'donor_dna_promoter_components'},\n",
       " {'table_name': u'entrez_aliases'},\n",
       " {'table_name': u'ephys_stimulus_set_groups'},\n",
       " {'table_name': u'ensembl_to_ncbi_mappings'},\n",
       " {'table_name': u'donor_dna_recombinases'},\n",
       " {'table_name': u'donor_dna_elements'},\n",
       " {'table_name': u'ephys_specimen_roi_plans'},\n",
       " {'table_name': u'ephys_pipette_types'},\n",
       " {'table_name': u'engineering_steps'},\n",
       " {'table_name': u'cell_reporters'},\n",
       " {'table_name': u'ephys_neural_tissue_templates_ephys_roi_plans'},\n",
       " {'table_name': u'donor_dna_recombinase_sites'},\n",
       " {'table_name': u'engineering_step_validations'},\n",
       " {'table_name': u'ephys_sweep_tags'},\n",
       " {'table_name': u'donor_dna_promoters_donor_dna_recombinase_sites'},\n",
       " {'table_name': u'facs_plate_templates'},\n",
       " {'table_name': u'genome_loci'},\n",
       " {'table_name': u'ercc_stock_pools'},\n",
       " {'table_name': u'file_sources'},\n",
       " {'table_name': u'gene_classifications'},\n",
       " {'table_name': u'genes_genotypes'},\n",
       " {'table_name': u'go_files'},\n",
       " {'table_name': u'genders'},\n",
       " {'table_name': u'go_file_types'},\n",
       " {'table_name': u'facs_resources'},\n",
       " {'table_name': u'external_controls'},\n",
       " {'table_name': u'flips'},\n",
       " {'table_name': u'header_elements'},\n",
       " {'table_name': u'ercc_diluted_pools'},\n",
       " {'table_name': u'mechanical_dissociation_types'},\n",
       " {'table_name': u'injection_materials_injection_templates'},\n",
       " {'table_name': u'ish_quality_metrics'},\n",
       " {'table_name': u'locations'},\n",
       " {'table_name': u'informatics_ranking_query_types'},\n",
       " {'table_name': u'injection_qualities'},\n",
       " {'table_name': u'job_transitions_plane_of_sections'},\n",
       " {'table_name': u'nucleases'},\n",
       " {'table_name': u'job_transitions_ages'},\n",
       " {'table_name': u'list_configurations'},\n",
       " {'table_name': u'injections_structures'},\n",
       " {'table_name': u'microarray_chipsets'},\n",
       " {'table_name': u'injections'},\n",
       " {'table_name': u'human_chrs_map'},\n",
       " {'table_name': u'pcrs'},\n",
       " {'table_name': u'injection_materials_injections'},\n",
       " {'table_name': u'job_transitions_structures'},\n",
       " {'table_name': u'pcr_names'},\n",
       " {'table_name': u'neuronal_models'},\n",
       " {'table_name': u'probe_types'},\n",
       " {'table_name': u'registration_points'},\n",
       " {'table_name': u'rna_amplifications'},\n",
       " {'table_name': u'rs_alignment_runs'},\n",
       " {'table_name': u'rs_tubesets'},\n",
       " {'table_name': u'neuronal_model_template_types'},\n",
       " {'table_name': u'plate_types'},\n",
       " {'table_name': u'rs_tubes'},\n",
       " {'table_name': u'rseq_oligos'},\n",
       " {'table_name': u'probe_sources'},\n",
       " {'table_name': u'rna_seq_experiments'},\n",
       " {'table_name': u'neuronal_model_templates_well_known_files'},\n",
       " {'table_name': u'nuclease_activities_talen_pairs'},\n",
       " {'table_name': u'pcrs_southern_blots'},\n",
       " {'table_name': u'reprogramming_methods'},\n",
       " {'table_name': u'publishes'},\n",
       " {'table_name': u'rna_amplification_sets'},\n",
       " {'table_name': u'sub_image_annotation_tags'},\n",
       " {'table_name': u'specimen_types'},\n",
       " {'table_name': u'rseq_tube_sets'},\n",
       " {'table_name': u'sub_image_annotations'},\n",
       " {'table_name': u'stimulus_sets'},\n",
       " {'table_name': u'stereotaxic_injections_structures'},\n",
       " {'table_name': u'structure_unionizes'},\n",
       " {'table_name': u'sub_image_annotation_tags_sub_image_annotations'},\n",
       " {'table_name': u'rna_amplification_inputs'},\n",
       " {'table_name': u'specimen_sets'},\n",
       " {'table_name': u'rseq_size_selection_inputs'},\n",
       " {'table_name': u'sub_image_types'},\n",
       " {'table_name': u'structure_sets_structures'},\n",
       " {'table_name': u'run_types'},\n",
       " {'table_name': u'cell_banks'},\n",
       " {'table_name': u'chemical_instances'},\n",
       " {'table_name': u'donor_medical_conditions'},\n",
       " {'table_name': u'talen_pairs_engineering_steps'},\n",
       " {'table_name': u'tissuecyte_calibrations'},\n",
       " {'table_name': u'transfection_methods'},\n",
       " {'table_name': u'transgenic_line_sources'},\n",
       " {'table_name': u'wash_buffer_types'},\n",
       " {'table_name': u'well_types'},\n",
       " {'table_name': u'plates'},\n",
       " {'table_name': u'tumor_features'},\n",
       " {'table_name': u'transgenic_induction_methods'},\n",
       " {'table_name': u'washes'},\n",
       " {'table_name': u'top_levels'},\n",
       " {'table_name': u'workflows'},\n",
       " {'table_name': u'talens'},\n",
       " {'table_name': u'cell_prep_specimen_roi_plans'},\n",
       " {'table_name': u'donor_dna_construct_designs'},\n",
       " {'table_name': u'neuron_reconstruction_tags_neuron_reconstructions'},\n",
       " {'table_name': u'reference_genomes'},\n",
       " {'table_name': u'stem_cell_lines'},\n",
       " {'table_name': u'wells'},\n",
       " {'table_name': u'annotationuser'},\n",
       " {'table_name': u'annotationgroup'},\n",
       " {'table_name': u'classificationitem'},\n",
       " {'table_name': u'annotatedvaluedefinition'},\n",
       " {'table_name': u'annotationproject'},\n",
       " {'table_name': u'classificationitemretired'},\n",
       " {'table_name': u'region'},\n",
       " {'table_name': u'targetentitytype'},\n",
       " {'table_name': u'annotationspace'},\n",
       " {'table_name': u'executables_projects'},\n",
       " {'table_name': u'genomicspreadsheet'},\n",
       " {'table_name': u'allseries4_6'},\n",
       " {'table_name': u'classificationannotation'},\n",
       " {'table_name': u'allseries4_6_2'},\n",
       " {'table_name': u'floatannotation'},\n",
       " {'table_name': u'annotationitem'},\n",
       " {'table_name': u'annotationsetupdates'},\n",
       " {'table_name': u'integerannotation'},\n",
       " {'table_name': u'annotationitemupdates'},\n",
       " {'table_name': u'annotationprojectuser'},\n",
       " {'table_name': u'planregionmapcell'},\n",
       " {'table_name': u'correcttargetentityvalues'},\n",
       " {'table_name': u'groupgroup'},\n",
       " {'table_name': u'regionregion'},\n",
       " {'table_name': u'revisionhistory'},\n",
       " {'table_name': u'ages_executables'},\n",
       " {'table_name': u'archives'},\n",
       " {'table_name': u'death_causes'},\n",
       " {'table_name': u'blocks_section_arrangements'},\n",
       " {'table_name': u'slides'},\n",
       " {'table_name': u'chromosomes'},\n",
       " {'table_name': u'vertexset'},\n",
       " {'table_name': u'imageseriesgene'},\n",
       " {'table_name': u'probe_template_batches'},\n",
       " {'table_name': u'access_rights'},\n",
       " {'table_name': u'avg_replicates'},\n",
       " {'table_name': u'cell_prep_tubes'},\n",
       " {'table_name': u'run_groups'},\n",
       " {'table_name': u'runplans'},\n",
       " {'table_name': u'cell_soma_locations'},\n",
       " {'table_name': u'blast_runs'},\n",
       " {'table_name': u'calculatedilutiontemplate'},\n",
       " {'table_name': u'audits'},\n",
       " {'table_name': u'death_manners'},\n",
       " {'table_name': u'cloning_vectors'},\n",
       " {'table_name': u'probes'},\n",
       " {'table_name': u'ages_job_transitions'},\n",
       " {'table_name': u'ages'},\n",
       " {'table_name': u'transcript_versions'},\n",
       " {'table_name': u'differentiations'},\n",
       " {'table_name': u'donor_dna_promoters'},\n",
       " {'table_name': u'down_sample_job_details'},\n",
       " {'table_name': u'drugs'},\n",
       " {'table_name': u'equalizations'},\n",
       " {'table_name': u'treatments'},\n",
       " {'table_name': u'erccs'},\n",
       " {'table_name': u'engineering_steps_helper_plasmids'},\n",
       " {'table_name': u'gene_symbol_aliases'},\n",
       " {'table_name': u'engineering_step_component_assemblies'},\n",
       " {'table_name': u'image_series'},\n",
       " {'table_name': u'plane_of_sections'},\n",
       " {'table_name': u'run_group_types'},\n",
       " {'table_name': u'runplans_treatments'},\n",
       " {'table_name': u'donor_dna_constructs'},\n",
       " {'table_name': u'projects'},\n",
       " {'table_name': u'executables_treatments'},\n",
       " {'table_name': u'equipment_classes'},\n",
       " {'table_name': u'homologene_groups'},\n",
       " {'table_name': u'ephys_stimulus_types'},\n",
       " {'table_name': u'donor_dna_recombinases_engineering_steps'},\n",
       " {'table_name': u'drug_instances'},\n",
       " {'table_name': u'education_levels'},\n",
       " {'table_name': u'engineering_step_component_designs'},\n",
       " {'table_name': u'molbio_authorities'},\n",
       " {'table_name': u'heatmap_types'},\n",
       " {'table_name': u'gene_tags'},\n",
       " {'table_name': u'gene_association_sources'},\n",
       " {'table_name': u'helper_plasmids'},\n",
       " {'table_name': u'gene_histories'},\n",
       " {'table_name': u'gene_classification_categories'},\n",
       " {'table_name': u'job_queues'},\n",
       " {'table_name': u'image_series_slides'},\n",
       " {'table_name': u'file_uploads'},\n",
       " {'table_name': u'image_series_tags'},\n",
       " {'table_name': u'job_transitions_treatments'},\n",
       " {'table_name': u'ldds'},\n",
       " {'table_name': u'neuronal_model_runs'},\n",
       " {'table_name': u'image_types'},\n",
       " {'table_name': u'isi_experiment_tags_isi_experiments'},\n",
       " {'table_name': u'job_transitions'},\n",
       " {'table_name': u'job_transitions_orientations'},\n",
       " {'table_name': u'gene_classification_uses'},\n",
       " {'table_name': u'genotypes'},\n",
       " {'table_name': u'job_transitions_projects'},\n",
       " {'table_name': u'go_files_specimens'},\n",
       " {'table_name': u'image_series_image_series_tags'},\n",
       " {'table_name': u'isi_experiment_tags'},\n",
       " {'table_name': u'jobs'},\n",
       " {'table_name': u'medical_conditions'},\n",
       " {'table_name': u'neuron_reconstructions'},\n",
       " {'table_name': u'frozen_media'},\n",
       " {'table_name': u'schema_migrations'},\n",
       " {'table_name': u'section_arrangements'},\n",
       " {'table_name': u'sections'},\n",
       " {'table_name': u'scans'},\n",
       " {'table_name': u'slide_groups'},\n",
       " {'table_name': u'rna_to_map_well_mappings'},\n",
       " {'table_name': u'rna_amplification_methods'},\n",
       " {'table_name': u'ophys_experiment_tags_ophys_experiments'},\n",
       " {'table_name': u'ontologies'},\n",
       " {'table_name': u'restriction_enzymes_southern_blots'},\n",
       " {'table_name': u'restriction_enzymes'},\n",
       " {'table_name': u'probe_template_types'},\n",
       " {'table_name': u'nuclease_activities'},\n",
       " {'table_name': u'resolutions'},\n",
       " {'table_name': u'rseq_size_selections'},\n",
       " {'table_name': u'probe_sets'},\n",
       " {'table_name': u'probe_results'},\n",
       " {'table_name': u'pcr_confirmation_sequencing_runs'},\n",
       " {'table_name': u'pcr_products'},\n",
       " {'table_name': u'plate_usages'},\n",
       " {'table_name': u'pcr_reaction_types'},\n",
       " {'table_name': u'rseq_size_selection_sets'},\n",
       " {'table_name': u'neuronal_model_templates'},\n",
       " {'table_name': u'orientations'},\n",
       " {'table_name': u'task_flow_steps'},\n",
       " {'table_name': u'refseq_homologenes'},\n",
       " {'table_name': u'races'},\n",
       " {'table_name': u'rna_seq_analysis_runs_rna_seq_experiments'},\n",
       " {'table_name': u'sub_image_qcstate_reasons'},\n",
       " {'table_name': u'structure_graphs'},\n",
       " {'table_name': u'normalization_groups'},\n",
       " {'table_name': u'differentiation_collection_days'},\n",
       " {'table_name': u'system_settings'},\n",
       " {'table_name': u'annotationplanvaluedefinition'},\n",
       " {'table_name': u'specimen_preparation_methods'},\n",
       " {'table_name': u'stimulus_sets_well_known_files'},\n",
       " {'table_name': u'southern_blot_label_types'},\n",
       " {'table_name': u'surgery_orientations'},\n",
       " {'table_name': u'slide_tags'},\n",
       " {'table_name': u'cell_samples'},\n",
       " {'table_name': u'blast_alignments'},\n",
       " {'table_name': u'task_flow_steps_task_templates'},\n",
       " {'table_name': u'gene_classification_sources'},\n",
       " {'table_name': u'ephys_features'},\n",
       " {'table_name': u'sequence_types'},\n",
       " {'table_name': u'rseq_library_prep_methods'},\n",
       " {'table_name': u'template_tags'},\n",
       " {'table_name': u'biospecimen_polygons'},\n",
       " {'table_name': u'facs_specimens'},\n",
       " {'table_name': u'job_states'},\n",
       " {'table_name': u'primer_reports'},\n",
       " {'table_name': u'slide_tags_slides'},\n",
       " {'table_name': u'source_roles'},\n",
       " {'table_name': u'structure_edges'}]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_query = \"SELECT TABLE_NAME FROM INFORMATION_SCHEMA.TABLES\"\n",
    "my_result = limsquery(my_query)\n",
    "my_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This is the basics of how a SQL query looks\n",
    "#SELECT TableA., TableB., TableC., TableD. FROM TableA JOIN TableB ON TableB.aID = TableA.aID \n",
    "#JOIN TableC ON TableC.cID = TableB.cID WHERE DATE(TableC.date)=date(now())\n",
    "#This allows us to pull in info from 4 different tables since everything we want is not \n",
    "#stored in one place, but there are overlaps to link things up. 10 digit IDs generally are your link"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ephys_roi_result_id</th>\n",
       "      <th>cell_id</th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13904</th>\n",
       "      <td>668475134</td>\n",
       "      <td>668475152</td>\n",
       "      <td>668475134</td>\n",
       "      <td>Slc17a6-IRES-Cre;Ai14-376996.04.01.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13905</th>\n",
       "      <td>668475449</td>\n",
       "      <td>668475465</td>\n",
       "      <td>668475449</td>\n",
       "      <td>Slc17a6-IRES-Cre;Ai14-376996.04.02.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13906</th>\n",
       "      <td>668489557</td>\n",
       "      <td>668489573</td>\n",
       "      <td>668489557</td>\n",
       "      <td>Ndnf-IRES2-dgCre;Ai14-378475.04.02.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13907</th>\n",
       "      <td>668531759</td>\n",
       "      <td>668531764</td>\n",
       "      <td>668531759</td>\n",
       "      <td>Ntsr1-Cre_GN220;Ai140;Sst-IRES-FlpO-377972-Ai6...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13908</th>\n",
       "      <td>668536310</td>\n",
       "      <td>668536315</td>\n",
       "      <td>668536310</td>\n",
       "      <td>Ntsr1-Cre_GN220;Ai140;Sst-IRES-FlpO-377972-Ai6...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ephys_roi_result_id    cell_id         id  \\\n",
       "13904            668475134  668475152  668475134   \n",
       "13905            668475449  668475465  668475449   \n",
       "13906            668489557  668489573  668489557   \n",
       "13907            668531759  668531764  668531759   \n",
       "13908            668536310  668536315  668536310   \n",
       "\n",
       "                                                    name  \n",
       "13904              Slc17a6-IRES-Cre;Ai14-376996.04.01.01  \n",
       "13905              Slc17a6-IRES-Cre;Ai14-376996.04.02.02  \n",
       "13906              Ndnf-IRES2-dgCre;Ai14-378475.04.02.01  \n",
       "13907  Ntsr1-Cre_GN220;Ai140;Sst-IRES-FlpO-377972-Ai6...  \n",
       "13908  Ntsr1-Cre_GN220;Ai140;Sst-IRES-FlpO-377972-Ai6...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# note both specimens and ephys roi results have an ID column, so we renamed the specimens column \n",
    "# as cell_id in the output\n",
    "\n",
    "lims_query = \"SELECT ephys_roi_results.id, specimens.id AS cell_id, specimens.name, specimens.ephys_roi_result_id \\\n",
    "FROM ephys_roi_results JOIN specimens ON specimens.ephys_roi_result_id = ephys_roi_results.id\"\n",
    "lims_df = get_lims_dataframe(lims_query)\n",
    "lims_df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This is the structure we want to follow for SQL:\n",
    "#SELECT your tables.columns\n",
    "#FROM these columns are coming if you are going to starting using\n",
    "#JOIN left ON right\n",
    "#WHERE gives you the ability to filter data\n",
    "#AND and OR add variability to your filters\n",
    "#\" opens and closes your query\"\n",
    "#\\ lets you wrap text\n",
    "#We can shorten ephys_roi_results to err and specimens to s, search for all of our features and use\n",
    "#JOIN to join the search results from ephys_roi_results and specimens following the outline from above"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_id</th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13904</th>\n",
       "      <td>668475152</td>\n",
       "      <td>668475134</td>\n",
       "      <td>Slc17a6-IRES-Cre;Ai14-376996.04.01.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13905</th>\n",
       "      <td>668475465</td>\n",
       "      <td>668475449</td>\n",
       "      <td>Slc17a6-IRES-Cre;Ai14-376996.04.02.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13906</th>\n",
       "      <td>668489573</td>\n",
       "      <td>668489557</td>\n",
       "      <td>Ndnf-IRES2-dgCre;Ai14-378475.04.02.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13907</th>\n",
       "      <td>668531764</td>\n",
       "      <td>668531759</td>\n",
       "      <td>Ntsr1-Cre_GN220;Ai140;Sst-IRES-FlpO-377972-Ai6...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13908</th>\n",
       "      <td>668536315</td>\n",
       "      <td>668536310</td>\n",
       "      <td>Ntsr1-Cre_GN220;Ai140;Sst-IRES-FlpO-377972-Ai6...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         cell_id         id                                               name\n",
       "13904  668475152  668475134              Slc17a6-IRES-Cre;Ai14-376996.04.01.01\n",
       "13905  668475465  668475449              Slc17a6-IRES-Cre;Ai14-376996.04.02.02\n",
       "13906  668489573  668489557              Ndnf-IRES2-dgCre;Ai14-378475.04.02.01\n",
       "13907  668531764  668531759  Ntsr1-Cre_GN220;Ai140;Sst-IRES-FlpO-377972-Ai6...\n",
       "13908  668536315  668536310  Ntsr1-Cre_GN220;Ai140;Sst-IRES-FlpO-377972-Ai6..."
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# callout the abreveations in the FROM section\n",
    "# ephys_roi_results err\n",
    "\n",
    "lims_query = \"SELECT err.id, s.id AS cell_id, s.name \\\n",
    "FROM ephys_roi_results err \\\n",
    "JOIN specimens s ON s.ephys_roi_result_id = err.id\"\n",
    "\n",
    "lims_df = get_lims_dataframe(lims_query)\n",
    "lims_df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>code</th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>908</th>\n",
       "      <td>H301</td>\n",
       "      <td>643584484</td>\n",
       "      <td>H17.06.013.12.04.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>909</th>\n",
       "      <td>H301</td>\n",
       "      <td>643589588</td>\n",
       "      <td>H17.06.013.12.04.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>910</th>\n",
       "      <td>H301</td>\n",
       "      <td>643601251</td>\n",
       "      <td>H17.06.013.12.13.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>911</th>\n",
       "      <td>H301</td>\n",
       "      <td>643619982</td>\n",
       "      <td>H17.06.013.12.13.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>912</th>\n",
       "      <td>H301</td>\n",
       "      <td>643625553</td>\n",
       "      <td>H17.06.013.12.13.03</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     code         id                 name\n",
       "908  H301  643584484  H17.06.013.12.04.01\n",
       "909  H301  643589588  H17.06.013.12.04.02\n",
       "910  H301  643601251  H17.06.013.12.13.01\n",
       "911  H301  643619982  H17.06.013.12.13.02\n",
       "912  H301  643625553  H17.06.013.12.13.03"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# callout the abreveations in the FROM section\n",
    "# ephys_roi_results err\n",
    "\n",
    "lims_query = \"SELECT err.id, s.id, s.name, proj.code \\\n",
    "FROM ephys_roi_results err \\\n",
    "JOIN specimens s ON s.ephys_roi_result_id = err.id \\\n",
    "JOIN projects proj ON s.project_id = proj.id \\\n",
    "WHERE proj.code = 'H301'\"\n",
    "\n",
    "lims_df = get_lims_dataframe(lims_query)\n",
    "lims_df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ephys_roi_result_id</th>\n",
       "      <th>code</th>\n",
       "      <th>recording_date</th>\n",
       "      <th>failed_bad_rs</th>\n",
       "      <th>workflow_state</th>\n",
       "      <th>name</th>\n",
       "      <th>failed_clogged_pipette</th>\n",
       "      <th>failed_other</th>\n",
       "      <th>project_id</th>\n",
       "      <th>failed_electrode_0</th>\n",
       "      <th>id</th>\n",
       "      <th>failed_no_seal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>601831626</td>\n",
       "      <td>T301</td>\n",
       "      <td>2017-07-10 21:03:11</td>\n",
       "      <td>True</td>\n",
       "      <td>manual_failed</td>\n",
       "      <td>Nos1-CreERT2;Ai14-333287.03.01.01</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>305094322</td>\n",
       "      <td>False</td>\n",
       "      <td>305094322</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>602660919</td>\n",
       "      <td>T301</td>\n",
       "      <td>2017-07-13 22:16:27</td>\n",
       "      <td>True</td>\n",
       "      <td>manual_failed</td>\n",
       "      <td>Scnn1a-Tg2-Cre;Ai14-333165.04.02.01</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>305094322</td>\n",
       "      <td>False</td>\n",
       "      <td>305094322</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>604693290</td>\n",
       "      <td>T301x</td>\n",
       "      <td>2017-07-26 22:50:46</td>\n",
       "      <td>True</td>\n",
       "      <td>manual_failed</td>\n",
       "      <td>Rbp4-Cre_KL100;Ai14-337193.04.02.01</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>300080300</td>\n",
       "      <td>False</td>\n",
       "      <td>300080300</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>605537141</td>\n",
       "      <td>T301x</td>\n",
       "      <td>2017-07-28 20:37:50</td>\n",
       "      <td>True</td>\n",
       "      <td>manual_failed</td>\n",
       "      <td>Gad2-IRES-Cre;Ai14-336420.04.02.01</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>300080300</td>\n",
       "      <td>False</td>\n",
       "      <td>300080300</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>605538385</td>\n",
       "      <td>T301x</td>\n",
       "      <td>2017-07-28 21:32:11</td>\n",
       "      <td>True</td>\n",
       "      <td>manual_failed</td>\n",
       "      <td>Gad2-IRES-Cre;Ai14-336420.02.02.01</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>300080300</td>\n",
       "      <td>False</td>\n",
       "      <td>300080300</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ephys_roi_result_id   code      recording_date  failed_bad_rs  \\\n",
       "67            601831626   T301 2017-07-10 21:03:11           True   \n",
       "68            602660919   T301 2017-07-13 22:16:27           True   \n",
       "69            604693290  T301x 2017-07-26 22:50:46           True   \n",
       "70            605537141  T301x 2017-07-28 20:37:50           True   \n",
       "71            605538385  T301x 2017-07-28 21:32:11           True   \n",
       "\n",
       "   workflow_state                                 name  \\\n",
       "67  manual_failed    Nos1-CreERT2;Ai14-333287.03.01.01   \n",
       "68  manual_failed  Scnn1a-Tg2-Cre;Ai14-333165.04.02.01   \n",
       "69  manual_failed  Rbp4-Cre_KL100;Ai14-337193.04.02.01   \n",
       "70  manual_failed   Gad2-IRES-Cre;Ai14-336420.04.02.01   \n",
       "71  manual_failed   Gad2-IRES-Cre;Ai14-336420.02.02.01   \n",
       "\n",
       "    failed_clogged_pipette  failed_other  project_id  failed_electrode_0  \\\n",
       "67                   False         False   305094322               False   \n",
       "68                   False         False   305094322               False   \n",
       "69                   False         False   300080300               False   \n",
       "70                   False         False   300080300               False   \n",
       "71                   False         False   300080300               False   \n",
       "\n",
       "           id  failed_no_seal  \n",
       "67  305094322           False  \n",
       "68  305094322           False  \n",
       "69  300080300           False  \n",
       "70  300080300           False  \n",
       "71  300080300           False  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#filters for cells that only failed due to access resistance over 20 MOhm\n",
    "#grabs the ephys roi result id, project code, recording date, cell name, cell id, and qc results\n",
    "\n",
    "lims_query = \"SELECT err.id, err.recording_date, err.failed_bad_rs, \\\n",
    "err.failed_electrode_0, err.failed_clogged_pipette, err.failed_no_seal, \\\n",
    "err.failed_other, err.workflow_state, s.name, s.ephys_roi_result_id, s.project_id, proj.id, proj.code \\\n",
    "FROM ephys_roi_results err JOIN specimens s ON s.ephys_roi_result_id = err.id \\\n",
    "JOIN projects proj ON s.project_id = proj.id \\\n",
    "WHERE (failed_bad_rs = 'TRUE' \\\n",
    "AND err.failed_electrode_0 = 'FALSE' \\\n",
    "AND err.failed_clogged_pipette = 'FALSE' \\\n",
    "AND err.failed_no_seal = 'FALSE' \\\n",
    "AND err.failed_other = 'FALSE' \\\n",
    "AND err.workflow_state != 'manual_passed') \\\n",
    "AND (proj.code = 'T301' OR proj.code = 'T301x')\"\n",
    "\n",
    "lims_df = get_lims_dataframe(lims_query)\n",
    "lims_df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_depth</th>\n",
       "      <th>ephys_roi_result_id</th>\n",
       "      <th>siv_default_ontology_id</th>\n",
       "      <th>scheduled_date</th>\n",
       "      <th>parent_y_coord</th>\n",
       "      <th>reference_space_id</th>\n",
       "      <th>task_flow_id</th>\n",
       "      <th>updated_at</th>\n",
       "      <th>requested_date_range_start</th>\n",
       "      <th>carousel_well_name</th>\n",
       "      <th>...</th>\n",
       "      <th>tissue_processing_id</th>\n",
       "      <th>ephys_qc_result</th>\n",
       "      <th>requested_by_id</th>\n",
       "      <th>y_coord</th>\n",
       "      <th>specimen_set_id</th>\n",
       "      <th>cell_reporter_id</th>\n",
       "      <th>login</th>\n",
       "      <th>pinned_radius</th>\n",
       "      <th>external_specimen_name</th>\n",
       "      <th>parent_x_coord</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9.0</td>\n",
       "      <td>None</td>\n",
       "      <td>2018-02-26 21:36:10.121126</td>\n",
       "      <td>2018-02-20 08:00:00</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>649368583</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>nadiad</td>\n",
       "      <td>None</td>\n",
       "      <td>360991</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9.0</td>\n",
       "      <td>None</td>\n",
       "      <td>2018-02-28 19:59:09.996577</td>\n",
       "      <td>2018-02-20 08:00:00</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>649368583</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>nadiad</td>\n",
       "      <td>None</td>\n",
       "      <td>360906</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9.0</td>\n",
       "      <td>None</td>\n",
       "      <td>2018-03-01 19:15:01.789758</td>\n",
       "      <td>2018-02-20 08:00:00</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>649368583</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>nadiad</td>\n",
       "      <td>None</td>\n",
       "      <td>360566</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>2018-02-20 20:02:02.656193</td>\n",
       "      <td>2018-02-20 08:00:00</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>649368583</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>nadiad</td>\n",
       "      <td>None</td>\n",
       "      <td>357433</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>2018-02-20 20:01:24.158513</td>\n",
       "      <td>2018-02-20 08:00:00</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>649368583</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>nadiad</td>\n",
       "      <td>None</td>\n",
       "      <td>356797</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 76 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  cell_depth ephys_roi_result_id siv_default_ontology_id scheduled_date  \\\n",
       "0       None                None                    None           None   \n",
       "1       None                None                    None           None   \n",
       "2       None                None                    None           None   \n",
       "3       None                None                    None           None   \n",
       "4       None                None                    None           None   \n",
       "\n",
       "  parent_y_coord  reference_space_id task_flow_id                 updated_at  \\\n",
       "0           None                 9.0         None 2018-02-26 21:36:10.121126   \n",
       "1           None                 9.0         None 2018-02-28 19:59:09.996577   \n",
       "2           None                 9.0         None 2018-03-01 19:15:01.789758   \n",
       "3           None                 NaN         None 2018-02-20 20:02:02.656193   \n",
       "4           None                 NaN         None 2018-02-20 20:01:24.158513   \n",
       "\n",
       "  requested_date_range_start carousel_well_name      ...        \\\n",
       "0        2018-02-20 08:00:00               None      ...         \n",
       "1        2018-02-20 08:00:00               None      ...         \n",
       "2        2018-02-20 08:00:00               None      ...         \n",
       "3        2018-02-20 08:00:00               None      ...         \n",
       "4        2018-02-20 08:00:00               None      ...         \n",
       "\n",
       "   tissue_processing_id  ephys_qc_result requested_by_id y_coord  \\\n",
       "0                  None             None       649368583    None   \n",
       "1                  None             None       649368583    None   \n",
       "2                  None             None       649368583    None   \n",
       "3                  None             None       649368583    None   \n",
       "4                  None             None       649368583    None   \n",
       "\n",
       "  specimen_set_id cell_reporter_id   login  pinned_radius  \\\n",
       "0            None             None  nadiad           None   \n",
       "1            None             None  nadiad           None   \n",
       "2            None             None  nadiad           None   \n",
       "3            None             None  nadiad           None   \n",
       "4            None             None  nadiad           None   \n",
       "\n",
       "  external_specimen_name parent_x_coord  \n",
       "0                 360991           None  \n",
       "1                 360906           None  \n",
       "2                 360566           None  \n",
       "3                 357433           None  \n",
       "4                 356797           None  \n",
       "\n",
       "[5 rows x 76 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT * \\\n",
    "FROM scheduled_procedures \\\n",
    "JOIN users ON users.id = scheduled_procedures.assigned_to_id \\\n",
    "JOIN scheduled_procedures_subjects sps ON scheduled_procedures.id = sps.scheduled_procedure_id \\\n",
    "JOIN specimens ON sps.subject_id = specimens.id \\\n",
    "WHERE users.login = 'nadiad' and scheduled_procedures.lab_request_status ='active' \\\n",
    "ORDER BY scheduled_procedures.created_at\"\n",
    "df = get_lims_dataframe(query)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>task_type</th>\n",
       "      <th>lab_request_status</th>\n",
       "      <th>task_id</th>\n",
       "      <th>department_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>mouse_name</th>\n",
       "      <th>updated_at</th>\n",
       "      <th>requested_date_range_start</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>batch_name</th>\n",
       "      <th>assigned_to_id</th>\n",
       "      <th>requested_by_id</th>\n",
       "      <th>comments</th>\n",
       "      <th>scheduled_date</th>\n",
       "      <th>methods_versions_map</th>\n",
       "      <th>task_flow_step_id</th>\n",
       "      <th>project_id</th>\n",
       "      <th>location_id</th>\n",
       "      <th>id</th>\n",
       "      <th>requested_date_range_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Task</td>\n",
       "      <td>active</td>\n",
       "      <td>665596389</td>\n",
       "      <td>None</td>\n",
       "      <td>2018-02-20 20:13:13.133917</td>\n",
       "      <td>Slc17a7-IRES2-Cre;Camk2a-tTA;Ai93-360991</td>\n",
       "      <td>2018-02-20 20:13:13.133917</td>\n",
       "      <td>2018-02-20 08:00:00</td>\n",
       "      <td>649497400</td>\n",
       "      <td>dillanb_C600_Brain Observatory Visual Coding_2...</td>\n",
       "      <td>186</td>\n",
       "      <td>649368583</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{}</td>\n",
       "      <td>563649249</td>\n",
       "      <td>305493902</td>\n",
       "      <td>None</td>\n",
       "      <td>665596392</td>\n",
       "      <td>2018-02-20 08:00:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Task</td>\n",
       "      <td>active</td>\n",
       "      <td>665596637</td>\n",
       "      <td>None</td>\n",
       "      <td>2018-02-20 20:13:17.837289</td>\n",
       "      <td>Vip-IRES-Cre;Ai148-360906</td>\n",
       "      <td>2018-02-20 20:13:17.837289</td>\n",
       "      <td>2018-02-20 08:00:00</td>\n",
       "      <td>650072655</td>\n",
       "      <td>dillanb_C600_Brain Observatory Visual Coding_2...</td>\n",
       "      <td>186</td>\n",
       "      <td>649368583</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{}</td>\n",
       "      <td>563649249</td>\n",
       "      <td>305493902</td>\n",
       "      <td>None</td>\n",
       "      <td>665596646</td>\n",
       "      <td>2018-02-20 08:00:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Task</td>\n",
       "      <td>active</td>\n",
       "      <td>665596826</td>\n",
       "      <td>None</td>\n",
       "      <td>2018-02-20 20:13:22.027657</td>\n",
       "      <td>Cux2-CreERT2;Camk2a-tTA;Ai93-360566</td>\n",
       "      <td>2018-02-20 20:13:22.027657</td>\n",
       "      <td>2018-02-20 08:00:00</td>\n",
       "      <td>649502537</td>\n",
       "      <td>dillanb_C600_Brain Observatory Visual Coding_2...</td>\n",
       "      <td>186</td>\n",
       "      <td>649368583</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{}</td>\n",
       "      <td>563649249</td>\n",
       "      <td>305493902</td>\n",
       "      <td>None</td>\n",
       "      <td>665596829</td>\n",
       "      <td>2018-02-20 08:00:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Task</td>\n",
       "      <td>active</td>\n",
       "      <td>665596853</td>\n",
       "      <td>None</td>\n",
       "      <td>2018-02-20 20:13:27.512618</td>\n",
       "      <td>Vip-IRES-Cre;Ai148-357433</td>\n",
       "      <td>2018-02-20 20:13:27.512618</td>\n",
       "      <td>2018-02-20 08:00:00</td>\n",
       "      <td>646205980</td>\n",
       "      <td>dillanb_C600_Brain Observatory Visual Coding_2...</td>\n",
       "      <td>186</td>\n",
       "      <td>649368583</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{}</td>\n",
       "      <td>563649249</td>\n",
       "      <td>305493902</td>\n",
       "      <td>None</td>\n",
       "      <td>665596856</td>\n",
       "      <td>2018-02-20 08:00:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Task</td>\n",
       "      <td>active</td>\n",
       "      <td>665597226</td>\n",
       "      <td>None</td>\n",
       "      <td>2018-02-20 20:13:38.828488</td>\n",
       "      <td>Slc17a7-IRES2-Cre;Camk2a-tTA;Ai93-356797</td>\n",
       "      <td>2018-02-20 20:13:38.828488</td>\n",
       "      <td>2018-02-20 08:00:00</td>\n",
       "      <td>646498173</td>\n",
       "      <td>dillanb_C600_Brain Observatory Visual Coding_2...</td>\n",
       "      <td>186</td>\n",
       "      <td>649368583</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{}</td>\n",
       "      <td>563649249</td>\n",
       "      <td>305493902</td>\n",
       "      <td>None</td>\n",
       "      <td>665597231</td>\n",
       "      <td>2018-02-20 08:00:00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  task_type lab_request_status    task_id department_id  \\\n",
       "0      Task             active  665596389          None   \n",
       "1      Task             active  665596637          None   \n",
       "2      Task             active  665596826          None   \n",
       "3      Task             active  665596853          None   \n",
       "4      Task             active  665597226          None   \n",
       "\n",
       "                  created_at                                mouse_name  \\\n",
       "0 2018-02-20 20:13:13.133917  Slc17a7-IRES2-Cre;Camk2a-tTA;Ai93-360991   \n",
       "1 2018-02-20 20:13:17.837289                 Vip-IRES-Cre;Ai148-360906   \n",
       "2 2018-02-20 20:13:22.027657       Cux2-CreERT2;Camk2a-tTA;Ai93-360566   \n",
       "3 2018-02-20 20:13:27.512618                 Vip-IRES-Cre;Ai148-357433   \n",
       "4 2018-02-20 20:13:38.828488  Slc17a7-IRES2-Cre;Camk2a-tTA;Ai93-356797   \n",
       "\n",
       "                  updated_at requested_date_range_start  subject_id  \\\n",
       "0 2018-02-20 20:13:13.133917        2018-02-20 08:00:00   649497400   \n",
       "1 2018-02-20 20:13:17.837289        2018-02-20 08:00:00   650072655   \n",
       "2 2018-02-20 20:13:22.027657        2018-02-20 08:00:00   649502537   \n",
       "3 2018-02-20 20:13:27.512618        2018-02-20 08:00:00   646205980   \n",
       "4 2018-02-20 20:13:38.828488        2018-02-20 08:00:00   646498173   \n",
       "\n",
       "                                          batch_name  assigned_to_id  \\\n",
       "0  dillanb_C600_Brain Observatory Visual Coding_2...             186   \n",
       "1  dillanb_C600_Brain Observatory Visual Coding_2...             186   \n",
       "2  dillanb_C600_Brain Observatory Visual Coding_2...             186   \n",
       "3  dillanb_C600_Brain Observatory Visual Coding_2...             186   \n",
       "4  dillanb_C600_Brain Observatory Visual Coding_2...             186   \n",
       "\n",
       "   requested_by_id comments scheduled_date methods_versions_map  \\\n",
       "0        649368583     None           None                   {}   \n",
       "1        649368583     None           None                   {}   \n",
       "2        649368583     None           None                   {}   \n",
       "3        649368583     None           None                   {}   \n",
       "4        649368583     None           None                   {}   \n",
       "\n",
       "   task_flow_step_id  project_id location_id         id  \\\n",
       "0          563649249   305493902        None  665596392   \n",
       "1          563649249   305493902        None  665596646   \n",
       "2          563649249   305493902        None  665596829   \n",
       "3          563649249   305493902        None  665596856   \n",
       "4          563649249   305493902        None  665597231   \n",
       "\n",
       "  requested_date_range_end  \n",
       "0      2018-02-20 08:00:00  \n",
       "1      2018-02-20 08:00:00  \n",
       "2      2018-02-20 08:00:00  \n",
       "3      2018-02-20 08:00:00  \n",
       "4      2018-02-20 08:00:00  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT scheduled_procedures.*, sps.subject_id, specimens.name AS mouse_name \\\n",
    "FROM scheduled_procedures \\\n",
    "JOIN users ON users.id = scheduled_procedures.assigned_to_id \\\n",
    "JOIN scheduled_procedures_subjects sps ON scheduled_procedures.id = sps.scheduled_procedure_id \\\n",
    "JOIN specimens ON sps.subject_id = specimens.id \\\n",
    "WHERE users.login = 'nadiad' and scheduled_procedures.lab_request_status ='active'\"\n",
    "df = get_lims_dataframe(query)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subthreshold</th>\n",
       "      <th>cell_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>Oxtr-T2A-Cre;Ai14-351471.04.01.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>Slc32a1-IRES-Cre;Ai14-326812.04.02.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>Slc32a1-IRES-Cre;Ai14-305535.06.02.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>Gad2-IRES-Cre;Ai14-267341.07.02.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>Slc17a6-IRES-Cre;Ai14-309388.03.01.01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   subthreshold                              cell_name\n",
       "0             3      Oxtr-T2A-Cre;Ai14-351471.04.01.01\n",
       "1             0  Slc32a1-IRES-Cre;Ai14-326812.04.02.05\n",
       "2             0  Slc32a1-IRES-Cre;Ai14-305535.06.02.01\n",
       "3             0     Gad2-IRES-Cre;Ai14-267341.07.02.02\n",
       "4             0  Slc17a6-IRES-Cre;Ai14-309388.03.01.01"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Jim asked to generate a csv with the rows being each patch-seq cell and the columns being the number \n",
    "#of sweeps for each of those cells that are of a certain stim.description (name)\n",
    "\n",
    "query = \"SELECT cell.name AS cell_name, \\\n",
    "COUNT(CASE WHEN ephys_stimuli.description LIKE '%%subthreshold%%' THEN 1 ELSE NULL END) AS subthreshold \\\n",
    "FROM specimens cell JOIN ephys_sweeps ess ON cell.id = ess.specimen_id \\\n",
    "JOIN ephys_stimuli ON ess.ephys_stimulus_id = ephys_stimuli.id \\\n",
    "WHERE cell.patched_cell_container NOTNULL \\\n",
    "GROUP BY cell.name\"\n",
    "\n",
    "df2 = get_lims_dataframe(query)\n",
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>noise_1</th>\n",
       "      <th>ramp</th>\n",
       "      <th>shortsquare</th>\n",
       "      <th>chirp</th>\n",
       "      <th>noise_2</th>\n",
       "      <th>subthreshold</th>\n",
       "      <th>cap_check</th>\n",
       "      <th>longsquare</th>\n",
       "      <th>triple</th>\n",
       "      <th>cell_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>Oxtr-T2A-Cre;Ai14-351471.04.01.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>Slc32a1-IRES-Cre;Ai14-326812.04.02.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Slc32a1-IRES-Cre;Ai14-305535.06.02.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Gad2-IRES-Cre;Ai14-267341.07.02.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Slc17a6-IRES-Cre;Ai14-309388.03.01.01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   noise_1  ramp  shortsquare  chirp  noise_2  subthreshold  cap_check  \\\n",
       "0        0     2            5      0        0             3          1   \n",
       "1        0     0            0      0        0             0          0   \n",
       "2        0     1            3      0        0             0          0   \n",
       "3        0     3            7      0        0             0          0   \n",
       "4        0     1            3      0        0             0          0   \n",
       "\n",
       "   longsquare  triple                              cell_name  \n",
       "0           4       0      Oxtr-T2A-Cre;Ai14-351471.04.01.01  \n",
       "1           4       0  Slc32a1-IRES-Cre;Ai14-326812.04.02.05  \n",
       "2           1       0  Slc32a1-IRES-Cre;Ai14-305535.06.02.01  \n",
       "3           0       0     Gad2-IRES-Cre;Ai14-267341.07.02.02  \n",
       "4           0       0  Slc17a6-IRES-Cre;Ai14-309388.03.01.01  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"SELECT cell.name AS cell_name, \\\n",
    "COUNT(CASE WHEN ephys_stimuli.description LIKE '%%subthreshold%%' THEN 1 ELSE NULL END) AS subthreshold, \\\n",
    "COUNT(CASE WHEN ephys_stimuli.description LIKE '%%C1LSFINEST150112%%' THEN 1 ELSE NULL END) AS longsquare, \\\n",
    "COUNT(CASE WHEN ephys_stimuli.description LIKE '%%C1SSFINEST150112%%' THEN 1 ELSE NULL END) AS shortsquare, \\\n",
    "COUNT(CASE WHEN ephys_stimuli.description LIKE '%%C1RP25PR1S141203%%' THEN 1 ELSE NULL END) AS ramp, \\\n",
    "COUNT(CASE WHEN ephys_stimuli.description LIKE '%%Blip%%' THEN 1 ELSE NULL END) AS cap_check, \\\n",
    "COUNT(CASE WHEN ephys_stimuli.description LIKE '%%C2CHIRP171103%%' THEN 1 ELSE NULL END) AS chirp, \\\n",
    "COUNT(CASE WHEN ephys_stimuli.description LIKE '%%C1NSD1SHORT17110%%' THEN 1 ELSE NULL END) AS noise_1, \\\n",
    "COUNT(CASE WHEN ephys_stimuli.description LIKE '%%C1NSD2SHORT17110%%' THEN 1 ELSE NULL END) AS noise_2, \\\n",
    "COUNT(CASE WHEN ephys_stimuli.description LIKE '%%C2SSTRIPLE171103%%' THEN 1 ELSE NULL END) AS triple \\\n",
    "FROM specimens cell JOIN ephys_sweeps ess ON cell.id = ess.specimen_id \\\n",
    "JOIN ephys_stimuli ON ess.ephys_stimulus_id = ephys_stimuli.id \\\n",
    "WHERE cell.patched_cell_container NOTNULL \\\n",
    "GROUP BY cell.name\"\n",
    "\n",
    "df2 = get_lims_dataframe(query)\n",
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'CLOSE' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-26-9507d485ed6c>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mCLOSE\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m: name 'CLOSE' is not defined"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
